File: scripts/setup_database.py
#!/usr/bin/env python3
"""
Database Setup Script for Supabase
Executes SQL schema to create all required tables
"""

import os
import sys
import json
from pathlib import Path
from dotenv import load_dotenv

# Load environment variables from .env.local
env_path = Path(__file__).parent.parent / ".env.local"
load_dotenv(env_path)

SUPABASE_URL = os.getenv("NEXT_PUBLIC_SUPABASE_URL")
SUPABASE_KEY = os.getenv("NEXT_PUBLIC_SUPABASE_ANON_KEY")

if not SUPABASE_URL or not SUPABASE_KEY:
    print("❌ Error: Missing Supabase credentials in .env.local")
    print("   Required: NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY")
    sys.exit(1)

# Read SQL file
sql_path = Path(__file__).parent.parent / "Backup" / "SUPABASE_SCHEMA.sql"
if not sql_path.exists():
    print(f"❌ Error: SQL schema file not found at {sql_path}")
    sys.exit(1)

with open(sql_path, "r") as f:
    sql_content = f.read()

print("\n" + "="*70)
print("🗄️  SUPABASE DATABASE SETUP")
print("="*70)
print(f"\n📍 Project: {SUPABASE_URL}")
print(f"📋 Schema file: {sql_path.name}\n")

# Parse SQL statements
statements = [
    s.strip() + ";"
    for s in sql_content.split(";")
    if s.strip() and not s.strip().startswith("--")
]

print(f"📊 Found {len(statements)} SQL statements\n")
print("="*70)
print("\n✅ SETUP INSTRUCTIONS:\n")
print("Option 1: Automated Setup (if via API available)")
print("---")

try:
    import requests
    
    headers = {
        "Authorization": f"Bearer {SUPABASE_KEY}",
        "Content-Type": "application/json",
        "Prefer": "return=representation",
    }
    
    # Try to check if rpc function exists
    print("⏳ Checking Supabase connection...")
    
    # Try direct SQL execution via REST
    success_count = 0
    failed_count = 0
    
    for i, statement in enumerate(statements, 1):
        stmt_short = statement[:60] + "..." if len(statement) > 60 else statement
        print(f"[{i:2d}/{len(statements)}] {stmt_short:<70}", end=" ")
        
        try:
            # Try using Supabase REST API for SQL execution
            # Note: This requires specific setup and may not work with anon key
            response = requests.post(
                f"{SUPABASE_URL}/rest/v1/rpc/exec_sql",
                json={"sql": statement},
                headers=headers,
                timeout=10,
            )
            
            if response.status_code in [200, 201]:
                print("✅")
                success_count += 1
            else:
                print(f"⚠️  ({response.status_code})")
                failed_count += 1
        except Exception as e:
            print("❌")
            failed_count += 1
    
    print(f"\n📊 Results: {success_count} succeeded, {failed_count} failed\n")
    
    if failed_count == 0:
        print("🎉 Database setup completed successfully!")
        sys.exit(0)
    
except ImportError:
    print("⚠️  requests library not installed\n")
except Exception as e:
    print(f"⚠️  Automated setup failed: {e}\n")

# Fallback: Manual setup instructions
print("\n" + "="*70)
print("Option 2: Manual Setup (via Supabase Dashboard)")
print("="*70)
print("\n📝 Steps:\n")
print("1. Go to: https://app.supabase.com")
print("2. Select your project")
print("3. Go to SQL Editor → New Query")
print("4. Copy and paste the entire SQL schema below:")
print("\n" + "-"*70)
print(sql_content)
print("-"*70)
print("\n5. Click 'Run' to execute")
print("\n✅ Tables will be created automatically\n")

# Save to file for easy copying
output_path = Path(__file__).parent.parent / "DATABASE_SCHEMA.sql"
with open(output_path, "w") as f:
    f.write(sql_content)

print(f"💾 SQL schema saved to: DATABASE_SCHEMA.sql")
print(f"   You can open this file and copy its contents into Supabase\n")
print("="*70 + "\n")

File: scripts/setup-database.mjs
#!/usr/bin/env node

/**
 * Database Setup Script for Supabase
 * This script executes the SQL schema against Supabase PostgreSQL database
 * 
 * Usage: node scripts/setup-db.js
 * 
 * Note: Requires environment variables:
 * - NEXT_PUBLIC_SUPABASE_URL
 * - DATABASE_PASSWORD (for direct connection) or admin API key
 */

import fs from "fs"
import path from "path"
import { fileURLToPath } from "url"
import dotenv from "dotenv"

// Load environment variables
dotenv.config({ path: ".env.local" })

const __dirname = path.dirname(fileURLToPath(import.meta.url))

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY

if (!supabaseUrl || !supabaseAnonKey) {
    console.error("❌ Error: Missing Supabase credentials in .env.local")
    console.error("   Required: NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY")
    process.exit(1)
}

// Extract project ID from Supabase URL
const projectId = supabaseUrl.split("//")[1].split(".")[0]
console.log(`\n🔧 Setting up database for project: ${projectId}`)
console.log(`📍 Supabase URL: ${supabaseUrl}\n`)

// Read SQL file
const sqlPath = path.join(__dirname, "../Backup/SUPABASE_SCHEMA.sql")
if (!fs.existsSync(sqlPath)) {
    console.error(`❌ Error: SQL schema file not found at ${sqlPath}`)
    process.exit(1)
}

const sqlContent = fs.readFileSync(sqlPath, "utf-8")

console.log(`📋 SQL Schema Loaded\n`)
console.log(`═══════════════════════════════════════════════════════════════`)
console.log(`\n📝 Instructions to create tables in Supabase:\n`)
console.log(`1. Go to: ${supabaseUrl}/project/sql/new`)
console.log(`\n2. Copy and paste the following SQL into the editor:\n`)
console.log(`═══════════════════════════════════════════════════════════════\n`)
console.log(sqlContent)
console.log(`\n═══════════════════════════════════════════════════════════════`)
console.log(`\n3. Click "Run" button to execute\n`)
console.log(`4. If successful, all tables will be created\n`)

// Save SQL to a file for easy copying
const outputPath = path.join(__dirname, "../SETUP_SQL.sql")
fs.writeFileSync(outputPath, sqlContent)
console.log(`💾 SQL saved to: SETUP_SQL.sql\n`)
console.log(`   You can paste this file directly into Supabase SQL editor\n`)

console.log(`═══════════════════════════════════════════════════════════════\n`)
console.log(`⚡ Attempting automated setup via REST API...\n`)

// Try alternative: Use Supabase REST API with direct SQL execution
// This requires a service role key, which we'll try to use via the auth token

async function setupDatabase() {
    try {
        // Try to use Supabase Edge Functions or REST API
        const dbSetupUrl = `${supabaseUrl}/rest/v1/rpc/exec_sql`

        console.log(`⏳ Checking if direct SQL execution is available...`)

        const response = await fetch(dbSetupUrl, {
            method: "OPTIONS",
        })

        if (response.status === 404) {
            console.log(`ℹ️  Direct rpc method not available (this is normal)`)
            console.log(`   Please use the manual method above\n`)
            return false
        }

        console.log(`✅ RPC method is available, attempting to execute SQL...`)

        // Split SQL into individual statements
        const statements = sqlContent
            .split(";")
            .map((stmt) => stmt.trim())
            .filter((stmt) => stmt && !stmt.startsWith("--"))
            .map((stmt) => stmt + ";")

        console.log(`\n📊 Found ${statements.length} SQL statements\n`)

        let successCount = 0
        for (let i = 0; i < statements.length; i++) {
            const stmt = statements[i]
            const shortStmt = stmt.length > 50 ? stmt.substring(0, 50) + "..." : stmt

            process.stdout.write(`[${i + 1}/${statements.length}] ${shortStmt}... `)

            try {
                const res = await fetch(dbSetupUrl, {
                    method: "POST",
                    headers: {
                        Authorization: `Bearer ${supabaseAnonKey}`,
                        "Content-Type": "application/json",
                    },
                    body: JSON.stringify({ sql: stmt }),
                })

                if (res.ok) {
                    console.log(`✅`)
                    successCount++
                } else {
                    console.log(`⚠️  (${res.status})`)
                }
            } catch (err) {
                console.log(`❌`)
            }
        }

        console.log(`\n📊 Results: ${successCount}/${statements.length} executed`)
        return successCount === statements.length
    } catch (error) {
        console.log(`\n⚠️  Automated setup not available\n`)
        return false
    }
}

setupDatabase().then((success) => {
    if (!success) {
        console.log(`\n🔗 Alternative: Use Supabase CLI\n`)
        console.log(`   npm install -g supabase`)
        console.log(`   supabase login`)
        console.log(`   supabase db push`)
        console.log(`\n`)
    }
})
